[yokurang] Added a zipper implementation and integrated it into interactive_viewer.ml accompanied with unit tests and an mli file.

yokurang opened the pull request:
Solution to issues https://github.com/panglesd/diffcessible/issues/4 and https://github.com/panglesd/diffcessible/issues/6

Run 
```
dune clean && dune build
```

and test using 
```
dune exec -- diffcessible diff.example
```

One concern I have is that I am unsure whether my implementation of using the zipper accomplished the task of a more efficient experience for very large diff files. Is there a way for me to check this? I tried manually traversing a large diff file and comparing using my changes and the old changes, but perhaps my diff file wasn't large enough to feel the difference? 

Any recommendations or thought would be appreciated. Thank you !

On a further note, I added a helper function for navigating between `Next` and `Prev` patches. This navigation has a built in check to ensure that we only navigate `Next` or `Prev` if there are elements to navigate to. Note the if statements in the code snippet below. This is my solution to issue https://github.com/panglesd/diffcessible/issues/4

```
let navigate direction zipper_of_patches =
  let z = Lwd.peek zipper_of_patches in
  match direction with
  | `Next ->
      if List.length z.after > 0 then begin
        Lwd.set zipper_of_patches (next_zipper z);
        Lwd.set index_v ((Lwd.peek index_v) + 1)
      end
  | `Prev ->
      if List.length z.before > 0 then begin
        Lwd.set zipper_of_patches (prev_zipper z);
        Lwd.set index_v ((Lwd.peek index_v) - 1)
      end
```

@Julow @panglesd

panglesd reviewed (changes requested):
Thanks, impressive work!

I've left some comments in the code, but there is one more that I cannot leave in the code, since it is in an unmodified part... you aren't actually using the zipper!

The "performance issue" is at line 10. For instance, suppose you are showing the 10000th hunk and want to show the 10001th one. This line will fetch it from the list of patches using `List.nth_opt`, which takes 10001 step: you need to traverse through the first 10000 elements of the list.
The zipper is here for exactly this reason: you always have direct access to the previous and next elements.

Julow commented on lib/zipper.ml:
```diff
@@ -0,0 +1,92 @@
+(** This file contains the zipper implementation *)
+
+type 'a zipper = {
+  before: 'a list;  
+  current: 'a option;
+  after: 'a list;
+}
```
I agree, an zipper into an empty list is a weird thing to work with and complicates the data structure a lot.
I'd rather avoid zipping into empty lists at all, for example by having `zipper_of_patches` returns a `_ zipper option`.

yokurang replied:
Resolved. The function zipper_of_list raises a custom exception when applied to an empty list

```
  let zipper_of_list lst =
    match lst with
    | [] -> raise Empty_list
    | x :: xs -> { before = []; focus = x; after = xs }
```

panglesd commented:
Small additional comment: we should add it somewhere in the doc since it is not currently there, but we format the code with ocamlformat!

So you should install ocamlformat version 0.26.1 and run `dune fmt`.

yokurang commented on lib/zipper.mli:
```diff
@@ -0,0 +1,7 @@
+type 'a zipper = { before : 'a list; current : 'a option; after : 'a list; }
+val zipper_of_list : 'a list -> 'a zipper
+val test_zipper_of_list : (int list -> int zipper) -> bool
```
Resolved

yokurang commented on lib/zipper.mli:
```diff
@@ -0,0 +1,7 @@
+type 'a zipper = { before : 'a list; current : 'a option; after : 'a list; }
```
Resolved

yokurang commented on lib/interactive_viewer.ml:
```diff
@@ -25,10 +29,25 @@ let current_hunks patches =
+  let z = Lwd.peek zipper_of_patches in
+  match direction with
+  | `Next ->
+      if List.length z.after > 0 then begin
```
Hi @panglesdm, do you mind clarifying what you mean by managing the `index_v` value? As of now, the current implementation for `next` and `prev` does not traverse any further once we reach the edge of the zipper. I believe we can customise this branch of the match statement to raise exceptions or log something to the user. In this sense, doesn't it make `index_v` obsolete? Otherwise I am misunderstanding something.

```
  let next z =
    match z.after with
    | [] -> z
    | x :: xs -> { before = z.focus :: z.before; focus = x; after = xs }

  let prev z =
    match z.before with
    | [] -> z
    | x :: xs -> { before = xs; focus = x; after = z.focus :: z.after }
```

P.S.: I pushed my current progress to the PR if you would like to take a look.

panglesd replied:
Yes, sure!

So, the `index_v` value contains an integer, which is updated to contain the index of the operation we are currently looking at.
Previously, this was used to get the current operation (with `List.nth_opt`). This was changed by your PR, and you now only use it to check bounds to avoid going "previous" or "next" when you are at the bounds.
However:
- It is in general good to avoid having to check bounds for each call of `Zipper.prev` or `Zipper.next`,
- In this case, it is not really useful to check the bounds: in your last version, you do not modify the zipper in the "out of bounds" situation!

We could completely get rid of the `index` value, since as we do not have to check the bounds, we are not using this value.
For another PR, we could want to have this value to add "Operation x out of X" in the top bar. When this happen, we will want to have Zipper itself manage the `index` value, instead of updating it at each call of `Zipper.next/prev`.

I hope it is clearer, do let me know otherwise!

yokurang replied:
Resolved

panglesd commented on lib/zipper.ml:
```diff
@@ -0,0 +1,57 @@
+(** This file contains the zipper implementation *)
+
+module Zipper : sig
```
Why did you decide to remove the `.mli`?
It is more conventional to put the interface in the mli file than in the ml file in this way. It also avoids having a module `Zipper.Zipper`, which is not ideal!

panglesd replied:
No problem! Your contributions are of high quality, and you are here to learn :slightly_smiling_face:

yokurang replied:
Resolved

yokurang commented:
Hi @Julow and @panglesd, I replaced the use of lists with zippers and resolved all of the mentioned comments. I added a little bit of extra features as they were quick to implement like "Operation x of X" on the topline, count of additions and removals, and colours. 

Best
Alan

yokurang commented:
> Small additional comment: we should add it somewhere in the doc since it is not currently there, but we format the code with ocamlformat!
> 
> So you should install ocamlformat version 0.26.1 and run `dune fmt`.

Added this to the README PR at https://github.com/panglesd/diffcessible/pull/25

panglesd reviewed (commented):
Thanks for your improvements!
I left some more comments in the code.

However, some of your additions clashes with issues assigned to other applicants! (the "Compute and display number of added and removed lines in status bar" and the "Display current operation in status bar" issues). While I can understand that it can be frustrating not to be able to add those "quick wins", we need to have good first issues for everyone.
The "Display current operation in status bar" even has an open PR, by someone assigned this issue!
So, could you remove all changes related to these from this PR?
The display of "Operation x of X" could have been a good first issue as well, but since it was not opened and not assigned yet, let's leave it in this PR!

We will assign you a "good medium issue" instead of having you fix all good first issues, if that is ok for you!

yokurang commented on test/test_zipper.ml:
```diff
@@ -0,0 +1,68 @@
+    let lst = [ 1; 2; 3; 4; 5 ] in
+    let z = candidate lst in
+    let test_b0 =
+      Diffcessible.Zipper.get_focus z = 1
```
Resolved.

yokurang commented on lib/zipper.mli:
```diff
@@ -0,0 +1,18 @@
+type 'a t = {
```
I believe resolved...?

panglesd replied:
Not really :)

Let me try to explain: in OCaml, there is a clear distinction between the interface (or in other words, the API of a module) and its implementation. The former is in the `.mli`, and the latter in the `.ml`.

The API should contain everything you need to _use_ a module, but nothing more. In particular, in a `.mli` you have the name of a value, and it's type, but _how_ the value is implemented is hidden.
Having this separation is very nice, because it allows to change easily the implementation, without breaking any user. But for this to really work we need to not only hide the implementation of values, but also the one of types!
OCaml allows to define "abstract types": types whose definition is hidden:
```ocaml
(** dict.ml *)
type t = (int * string) list
let find dict key = ...

(** dict.mli *)
type t
val find : t -> int -> string
```

If we included the definition of the type `t` in the interface, some users could want to write code that depend on the fact that a dictionary is an association list, which is an implementation detail. And, if we want to change that to something else, say a Hashtable, then this would break all code relying on the implementation detail.

Since interface are the only thing that should be looked at when using a module, documentation comments should go there, and not in the `.ml`! For instance, the docs in ocaml.irg (such as [this one](https://ocaml.org/p/bos/latest/doc/Bos/OS/File/index.html)) are generated from `.mli`s that contain doc comments!

yokurang commented on lib/zipper.mli:
```diff
@@ -0,0 +1,18 @@
+
+exception Empty_list
+
+val zipper_of_list : 'a list -> 'a t
```
Resolved.

yokurang commented on lib/interactive_viewer.ml:
```diff
@@ -1,52 +1,139 @@
 open Nottui
 module W = Nottui_widgets
 open Lwd_infix
+open Zipper
```
Resolved.

yokurang commented:
Hi @panglesd I implemented all of your comments and I removed (I believe all) of the overlapping solutions to other issues that you mentioned. Please let me know if I missed something, and I will remove it from my PR. 

If possible, I'd like to take https://github.com/panglesd/diffcessible/issues/15 and its related issue https://github.com/panglesd/diffcessible/issues/14 as I think it'd be a good challenge. 

Thank you !

yokurang

panglesd commented on lib/interactive_viewer.ml:
```diff
@@ -2,85 +2,98 @@ open Nottui
+  let p = Zipper.get_focus z in
+  let hunks =
+    List.map
+      (fun h -> W.string ~attr:Notty.A.(fg lightblue) (string_of_hunk h))
```
Hmm, why did you decide to change the color of the hunks? :thinking:

yokurang replied:
I think it was a personal preference, but I'll remove it for now until we establish an official colour!

panglesd reviewed (approved):
That is very close to be merged! There is just one issue about color, and some comment on interface vs implementation that need to be resolved.

yokurang commented:
Thanks for the explanation @panglesd! I moved the comments to the .mli file and removed the colour from the current hunks function. Thanks for taking a look!

panglesd commented:
Great! It only remains to make the type in `zipper.mli` asbtract.

Here is another [resource](https://courses.cs.cornell.edu/cs3110/2021sp/textbook/modules/abstract_types.html) on abstract types that you can have a read!

yokurang commented:
@panglesd , sorry for the confusion, but I believe I've done it right this time.

panglesd commented:
Thanks a lot, impressive work!

Let's merge!
